cpp/src/gis/point_in_polygon.cu

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Christian Cordova Estrada <christianc@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gis.hpp"
+#include "point_in_polygon.hpp"
+
+#include <bitmask/BitMask.cuh>
+#include "bitmask/bitmask_ops.hpp"
+#include "utilities/type_dispatcher.hpp"
+
+#include <type_traits>
+
+namespace {
+
+    /** 
+ * @brief Compute the orientation of p3 from two others points p1 and p2
+ *
+ * @param[in] p1_x: Longitude of the first point p1
+ * @param[in] p1_y: Latitude of the first point p1
+ * @param[in] p2_x: Longitude of the second point p2
+ * @param[in] p2_y: Latitude of the second point p2
+ * @param[in] p3_x: Longitude of the third point p3
+ * @param[in] p3_y: Latitude of the third point p3
+ *
+ * @returns positive if it's clockwise, negative if is counter-clockwise and 0 if is colinear
+ */
+template <typename T>
+__device__ T orientation(const T p1_x, const T p1_y, const T p2_x, const T p2_y, const T p3_x, const T p3_y)
+{
+    return ((p2_y - p1_y) * (p3_x - p2_x) - (p2_x - p1_x) * (p3_y - p2_y));
+}
+
+/** 
+ *  @brief Determine whether or not coordinates (query points) are completely inside a static polygon
+ *
+ * @param[in] poly_lats: Pointer to latitudes of a polygon
+ * @param[in] poly_lons: Pointer to longitudes of a polygon
+ * @param[in] point_lats: Pointer to latitudes of many query points
+ * @param[in] point_lons: Pointer to longitudes of many query points
+ * @param[in] poly_size: Size of polygon
+ * @param[in] point_size: Total number of query points
+ * @param[out] point_is_in_polygon: Pointer indicating if the i-th query point is inside or not
+ *
+ * @returns
+ */
+ template <typename T>
+ __global__ void point_in_polygon_kernel(const T* poly_lats,
+                                    const T* poly_lons,
+                                    const T* point_lats,
+                                    const T* point_lons,
+                                    gdf_size_type poly_size,
+                                    gdf_size_type point_size,
+                                    int8_t* point_is_in_polygon)
```

Comment:
```suggestion
                                    cudf::bool8* const __restrict__ point_is_in_polygon)
```

Reply:
It should be:   `cudf::bool8* __restrict__  point_is_in_polygon`    without the `const` word because   const __restrict__   is optimized to read-only data. Right?

Reply:
So there's different places `const` can be applied for a pointer. I always read right to left to help me remember:

1. `int * const p` - `p` is a _constant_ pointer to an `int`
   - `*p = 10` this is fine
   - `p = &a` this is not allowed
2. `int const * p` - `p` is a pointer to a _constant_ `int`
   - `*p = 10` this is not allowed
   - `p = &a` this is fine
3. `int const * const p` - `p` is a _constant_ pointer to a _constant_ `int`
   - `*p = 10` this is not allowed
   - `p = &a` this is not allowed

So in your case, `point_is_in_polygon` is an output array, so you want to be able to modify the `cudf::bool8` elements, but you don't want to change what `point_is_in_polygon` is pointing to.

Making the _pointer_ be `const`, in addition to `__restrict__`, can sometimes enable the compiler to better optimize your code.